### PR TITLE
Remove `typing.Self` workaround

### LINKED ISSUE #9254: Remove `typing.Self` workaround
### Ensure stubs packages are not installed

- [X] No sqlalchemy stub packages is installed (both `sqlalchemy-stubs` and `sqlalchemy2-stubs` are not compatible with v2)

### Verify if the api is typed

- [X] The api is not in a module listed in [#6810](https://github.com/sqlalchemy/sqlalchemy/issues/6810) so it should pass type checking

### Describe the typing issue

New `mypy` version `1.0.0` supports [PEP 673](https://peps.python.org/pep-0673/) so we can replace workarounds that `slqalchemy` has.

### To Reproduce

```python
SelfInsert = typing.TypeVar("SelfInsert", bound="Insert")


class Insert(StandardInsert):
    def on_duplicate_key_update(self: SelfInsert, *args, **kw) -> SelfInsert:
```

### Error

```
# No error
``` 

### Versions

- OS: macOS
- Python: python3.11
- SQLAlchemy: 2.0.2
- Type checker (eg: mypy 0.991, pyright 1.1.290, etc): mypy 1.0.0

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by uriyyo**
Fixes: #9254

### Description

Remove `typing.Self` workaround. New `mypy` version 1.0.0 supports [PEP 673](https://peps.python.org/pep-0673/) so we can replace workarounds that `slqalchemy` has.

### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [x] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

### CODE REVIEW

**Comment by CaselIT** [comment]
Pyright is also happy with it, correct? ( Note that it's probable that main has errors on pyright)

File: lib/sqlalchemy/ext/asyncio/base.py [hunk 1]
```diff
@@ -73,38 +68,36 @@ def _assign_proxied(self, target: Optional[_PT]) -> Optional[_PT]:
...
 
     @overload
     @classmethod
     def _retrieve_proxy_for_target(
-        cls: Type[SelfReversibleProxy],
+        cls,
         target: _PT,
         regenerate: Literal[True] = ...,
-    ) -> SelfReversibleProxy:
+    ) -> Self:
```
Comment (CaselIT) [inline lib/sqlalchemy/ext/asyncio/base.py:87]: You can use it also on classmethods?

File: doc/build/changelog/unreleased_20/9255.rst [hunk 2]
```diff
@@ -0,0 +1,5 @@
+.. change::
+    :tags: typing
+    :tickets: 9254
+
+    Remove ``typing.Self`` workaround.
```
Comment (CaselIT) [inline doc/build/changelog/unreleased_20/9255.rst:5]: Feel free to add "pull request courtesy of .."

File: lib/sqlalchemy/sql/traversals.py [hunk 3]
```diff
@@ -211,8 +200,8 @@ class GenerativeOnTraversal(HasShallowCopy):
     __slots__ = ()
 
     def _generate(
-        self: SelfGenerativeOnTraversal,
-    ) -> SelfGenerativeOnTraversal:
+        self,
```
Comment (CaselIT) [inline lib/sqlalchemy/sql/traversals.py:203]: can you remove the tailing comma, so that black formats it inline?

**Review by CaselIT — COMMENTED** [review]
Great work, thanks for doing this.

I've left a couple of comments

File: lib/sqlalchemy/ext/asyncio/base.py (hunk 1 above)
Comment (uriyyo) [reply lib/sqlalchemy/ext/asyncio/base.py:87]: Yup, here is a link - https://peps.python.org/pep-0673/#use-in-classmethod-signatures

File: lib/sqlalchemy/sql/traversals.py (hunk 3 above)
Comment (uriyyo) [reply lib/sqlalchemy/sql/traversals.py]: Fixed, thanks for poiting on it

File: lib/sqlalchemy/ext/asyncio/base.py (hunk 1 above)
Comment (CaselIT) [reply lib/sqlalchemy/ext/asyncio/base.py:87]: nice.

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision f3f1eb9c6636f449172646d5d1e73e9020286230 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change f3f1eb9c6636f449172646d5d1e73e9020286230: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4432

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

thanks for doing this!  ill have a closer look soon.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4432

**Comment by zzzeek** [comment]
has some errors, is anyone looking into them ?

**Comment by CaselIT** [comment]
Strange that the github CI is happy with it. 

Oh it does `tox -e mypy ` not `tox -e pep484`.
It's so wired that github runs the `mypy` tox env only. I guess I forgot to add it

**Comment by CaselIT** [comment]
> Strange that the github CI is happy with it.
> 
> Oh it does `tox -e mypy ` not `tox -e pep484`. It's so wired that github runs the `mypy` tox env only. I guess I forgot to add it

Updated on 33dc651030b900bad31842df927fb8454b9a75d1. should run from the next PR

**Comment by uriyyo** [comment]
@CaselIT @zzzeek Errors have been fixed, please, review again

**Comment by CaselIT** [comment]
> > Strange that the github CI is happy with it.
> > Oh it does `tox -e mypy ` not `tox -e pep484`. It's so wired that github runs the `mypy` tox env only. I guess I forgot to add it
> 
> Updated on [33dc651](https://github.com/sqlalchemy/sqlalchemy/commit/33dc651030b900bad31842df927fb8454b9a75d1). should run from the next PR

oh cool, it added the new gate here too.

> @CaselIT @zzzeek Errors have been fixed, please, review again

Thanks, It seems ok from the gate now. let's re-add it in gerrit too

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 2947df8ada79f5c3afe9c838e65993302199c2f7 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset 2947df8ada79f5c3afe9c838e65993302199c2f7 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4432

**Comment by uriyyo** [comment]
@zzzeek Great solution, thanks for pinging me

**Pull request closed without merging**